bridge/bridge/deploy/local.py

```diff
@@ -0,0 +1,283 @@
+        if self.serve_handle.poll() is not None:
+            logger.info(
+                f"Detected stopped model deployment {self.model_version}."
+            )
+            destroy_failed_conda(self.model_uri)
+            self._install()
+            self.serve_handle = None
+
+    def _serve(self):
+        model_version = self.model_version
+        path = f"{model_version.model_name}:{model_version.version_id}"
+        stdout = open(f"{path}.stdout", "w")
+        stderr = open(f"{path}.stderr", "w")
+
+        env = {}
+        if env_path := os.environ.get("PATH"):
+            env["PATH"] = env_path
+        if env_home := os.environ.get("HOME"):
+            env["HOME"] = env_home
+
+        self.serve_handle = Popen(
+            [
+                "mlflow",
+                "models",
+                "serve",
+                "--model-uri",
+                self.model_uri,
+                "--port",
+                str(self.port),
+                "--host",
+                "127.0.0.1",
+            ],
+            stdout=stdout,
+            stderr=stderr,
+            preexec_fn=os.setsid,
+            env=env,
+        )
+
+    def destroy(self):
+        if self.serve_handle is not None:
+            os.killpg(os.getpgid(self.serve_handle.pid), signal.SIGTERM)
+            self.serve_handle.wait()
+            self.serve_handle = None
+
+
+class LocalDeployTarget(DeployTarget):
+    target_name = "local"
+
+    @property
+    def target_id(self) -> str:
+        return socket.gethostname()
+
+    def list_models(self) -> List[Model]:
+        # Run reconciliation loop for deployments
+        mvs = list(self.running_models.keys())
+        for mv in mvs:
+            self.running_models[mv].check()
+
+        result = []
+        for model_name, d in self.routes.items():
+            model = Model(name=model_name, versions={})
+            for stage, port in d.items():
+                for mv, md in self.running_models.items():
+                    if port == md.port:
+                        model.versions[stage] = set([mv])
+                        break
+            result.append(model)
+        logger.info(result)
+        return result
+
+    def teardown(self):
+        # TODO: Shutdown Flask
+        self.routes = {}
+        models = list(self.running_models.keys())
+        for model in models:
+            self.running_models[model].destroy()
+            del self.running_models[model]
+
+    def __init__(self):
+        self.running_models: Dict[ModelVersion, ModelDeployment] = {}
+        # model, stage, port
+        self.routes: Dict[str, Dict[str, int]] = {}
+        self.proxy = LocalDeploymentProxy(self)
+        self.proxy.run()
+
+    def init(self):
+        pass
+
+    def create_versions(self, new_versions: Dict[ModelVersion, Artifact]):
+        for model_version, artifact in new_versions.items():
+            if model_version in self.running_models:
+                logger.warning(
```

Comment:
Seems like info not warning. Having each version run only once is desriable/efficient behavior? (Assuming a world where one version can exist in many stages)

Reply:
This would indicate that the Models reported by the deployment target did not include this model, and the control loop instructed us to deploy it again. I think there is something mildly wrong if that happens.

Reply:
I see